/*
** Hand-e project, 2024
** hostasphere python profiler api
** File description:
** session.proto
*/

syntax = "proto3";

package token;

option go_package = "./proto";

service SessionService {
    rpc SaveSession (SaveSessionRequest) returns (SaveSessionResponse);
}

message UsageAtTime {
    double time = 1;
    double memory_usage = 2;
}

message TrackAnnotation {
    double time = 1;
    string annotation = 2;
    string color = 3; // hex color
}

message Session {
    string session_uuid = 1;
    string session_tag = 2;
    double start_time = 3;
    int64 start_date = 4;
    double end_time = 5;
    int64 end_date = 6;
    double execution_time = 7; // in milliseconds
    repeated UsageAtTime memory_usage = 8;
    repeated UsageAtTime cpu_usage = 9;
    repeated UsageAtTime disk_usage = 10;
    repeated UsageAtTime network_usage = 11;
    repeated UsageAtTime tokens_usage = 12;
    string token_id = 13;
    int64 pid = 14;
    string hostname = 15;
    string os = 16;
    string os_version = 17;
    string kernel_version = 18;
    string architecture = 19;
    string python_version = 20;
    string processor = 21;
    int32 cpu_count = 22;
    double boot_time = 23;
    string current_user = 24;
    repeated TrackAnnotation track_annotations = 25;
    int64 total_tokens = 26;
}

message SaveSessionRequest {
    string token = 1;
    Session session = 2;
}

message SaveSessionResponse {
    bool ok = 1;
    string message = 2;
}
